### PR TITLE
[SYSTEMDS-206] HDF5 File Format

### PR DESCRIPTION
**Pull request by fathollahzadeh**
This commit is an implementation for HDF5 data format files. The HDF5 The file format specification is available from the HDF Group [here](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) 

Currently, this patch is targeting the HDF5 format with supporting some features of the HDF5. The supported features:
 1. Two Dimension Data (Matrix)
 2. Matrix with FP64 (double) Data Type
3. Single Dataset
4. Single Group 
5. Contiguous Dataset
6. Read/Write and SystemDS Parallel Feature

### CODE REVIEW

File: src/main/java/org/apache/sysds/runtime/io/hdf5/H5ContiguousDataset.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.sysds.runtime.io.hdf5.message.H5DataTypeMessage;
+
+import java.nio.ByteBuffer;
+
+import static java.nio.ByteOrder.LITTLE_ENDIAN;
+
+public class H5ContiguousDataset {
+
+	private final H5RootObject rootObject;
+	//private final H5ObjectHeader objectHeader;
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/io/hdf5/H5ContiguousDataset.java:34]: remove unused code

File: src/main/java/org/apache/sysds/runtime/io/hdf5/message/H5Message.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		return message;
+	}
+
+	protected H5BufferBuilder toBuffer() {
+		H5BufferBuilder bb = new H5BufferBuilder();
+		this.toBuffer(bb);
+		return bb;
+	}
+
+	public void toBuffer(H5BufferBuilder bb) {
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/io/hdf5/message/H5Message.java:59]: should this method do nothing?

File: src/test/java/org/apache/sysds/test/functions/io/hdf5/ReadHDF5Test.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			String datasetName = "DATASET_1";
+
+			//TODO: add RScript for verify file format
+
+			fullDMLScriptName = HOME + getTestName() + "_" + testNumber + ".dml";
+			programArgs = new String[] {"-args", inputMatrixName, datasetName, output("Y")};
+
+			runTest(true, false, null, -1);
+
+			HashMap<MatrixValue.CellIndex, Double> YSYSTEMDS = readDMLMatrixFromOutputDir("Y");
```
Comment (Baunsgaard) [inline src/test/java/org/apache/sysds/test/functions/io/hdf5/ReadHDF5Test.java:80]: Missing verification of read values.

File: src/test/java/org/apache/sysds/test/functions/io/hdf5/WriteHDF5Test.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			//TODO: add RScript for verify file format
+
+			fullDMLScriptName = HOME + getTestName() + "_" + testNumber + ".dml";
+			programArgs = new String[] {"-args", input("A"), outMatrixName, datasetName};
+
+			//generate actual datasets
+			double[][] A = getRandomMatrix(rows, cols, 0, 10, 1, 714);
+			writeInputMatrixWithMTD("A", A, false);
+
+			runTest(true, false, null, -1);
```
Comment (Baunsgaard) [inline src/test/java/org/apache/sysds/test/functions/io/hdf5/WriteHDF5Test.java:91]: Missing verify of the written matrix.

**Comment by fathollahzadeh** [comment]
> A few comments, I have issues with the syntax for @,,, , double array allocations, and the tests. but otherwise it looks fine.
> I am unsure how much of the code is actually executed in the tests, maybe you want to see what parts are covered using jacoco?
> 
> most importantly the tests are missing verification of the actual values parsed and written to disk since the tests currently only verify that there are no crashes.

Thanks for reviewing PR @Baunsgaard . I really appreciate your review and your comments.

I fixed most of them and pushed them in separate commits. Also, some short responses to resolve the comments.

File: src/main/java/org/apache/sysds/parser/DMLTranslator.java [hunk 5]
```diff
@@ -1058,6 +1058,10 @@ public void constructHops(StatementBlock sb) {
 					case FEDERATED:
 						ae.setOutputParams(ae.getDim1(), ae.getDim2(), -1, ae.getUpdateType(), -1);
 						break;
+					case HDF5:
+						// write output in HDF5 format
+						ae.setOutputParams(ae.getDim1(), ae.getDim2(), ae.getNnz(), ae.getUpdateType(), ConfigurationManager.getBlocksize());
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/parser/DMLTranslator.java]: it was a wrong call. I fixed it in the commits.

File: src/main/java/org/apache/sysds/runtime/instructions/cp/VariableCPInstruction.java [hunk 6]
```diff
@@ -48,14 +48,7 @@
...
 import org.apache.sysds.runtime.instructions.InstructionUtils;
-import org.apache.sysds.runtime.io.FileFormatProperties;
-import org.apache.sysds.runtime.io.FileFormatPropertiesCSV;
-import org.apache.sysds.runtime.io.FileFormatPropertiesLIBSVM;
-import org.apache.sysds.runtime.io.IOUtilFunctions;
-import org.apache.sysds.runtime.io.ListReader;
-import org.apache.sysds.runtime.io.ListWriter;
-import org.apache.sysds.runtime.io.WriterMatrixMarket;
-import org.apache.sysds.runtime.io.WriterTextCSV;
+import org.apache.sysds.runtime.io.*;
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/instructions/cp/VariableCPInstruction.java]: fixed!

File: src/main/java/org/apache/sysds/runtime/io/MatrixWriterFactory.java [hunk 7]
```diff
@@ -77,6 +77,15 @@ public static MatrixWriter createMatrixWriter(FileFormat fmt, int replication, F
 					writer = new WriterBinaryBlock(replication);
 				break;
 
+			case HDF5:
+				if(props != null && !(props instanceof FileFormatPropertiesHDF5))
+					throw new DMLRuntimeException("Wrong type of file format properties for HDF5 writer.");
+				if( ConfigurationManager.getCompilerConfigFlag(ConfigType.PARALLEL_CP_WRITE_TEXTFORMATS) )
+					writer = new WriterHDF5Parallel((FileFormatPropertiesHDF5) props);
+				else
+					writer = new WriterHDF5((FileFormatPropertiesHDF5) props);
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/MatrixWriterFactory.java]: fixed!

File: src/main/java/org/apache/sysds/runtime/io/ReaderHDF5.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+			BufferedInputStream bis = new BufferedInputStream(fs.open(files.get(fileNo)));
+			lnnz += readMatrixFromHDF5(bis, datasetName, dest, row, rlen, clen, blen);
+		}
+		//post processing
+		dest.setNonZeros(lnnz);
+
+		return dest;
+	}
+
+	public static long readMatrixFromHDF5(BufferedInputStream bis, String datasetName, MatrixBlock dest,
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/ReaderHDF5.java:127]: It should be public. I called it in another class.

File: src/main/java/org/apache/sysds/runtime/io/ReaderHDF5Parallel.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+		//create and execute tasks
+		try {
+			ExecutorService pool = CommonThreadPool.get(_numThreads);
+			ArrayList<ReadHDF5Task> tasks = new ArrayList<>();
+			rlen = src.getNumRows();
+			int blklen = (int) Math.ceil((double) rlen / _numThreads);
+			for(int i = 0; i < _numThreads & i * blklen < rlen; i++) {
+				BufferedInputStream bis = new BufferedInputStream(fs.open(path));
+				MutableInt row = new MutableInt(i * blklen);
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/ReaderHDF5Parallel.java]: Yes, we don't need a MutableInt. I replaced it with primitive variable.

File: src/main/java/org/apache/sysds/runtime/io/hdf5/H5Constants.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysds.runtime.io.hdf5;
+
+public final class H5Constants {
+
+	private H5Constants() {
+		throw new AssertionError("No instances of Constants");
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/hdf5/H5Constants.java]: It was an extra contractor. I removed it.

File: src/main/java/org/apache/sysds/runtime/io/hdf5/H5ContiguousDataset.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+			e.printStackTrace();
+			throw new H5Exception("Failed to map data buffer for dataset", e);
+		}
+	}
+
+	public ByteBuffer getDataBuffer() {
+		try {
+			ByteBuffer data = rootObject.readBufferFromAddress(dataLayoutMessage.getAddress(),
+				(int) (this.dataSpaceMessage.getTotalLength() * this.dataTypeMessage.getDoubleDataType().getSize()));
+			data.order(LITTLE_ENDIAN);
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/hdf5/H5ContiguousDataset.java:51]: It was an extra method. Also, it was just for a test!

File: src/test/java/org/apache/sysds/test/functions/io/hdf5/ReadHDF5Test.java (hunk 3 above)
Comment (fathollahzadeh) [reply src/test/java/org/apache/sysds/test/functions/io/hdf5/ReadHDF5Test.java:84]: R scripts for read/write verification added

File: src/test/java/org/apache/sysds/test/functions/io/hdf5/WriteHDF5Test.java (hunk 4 above)
Comment (fathollahzadeh) [reply src/test/java/org/apache/sysds/test/functions/io/hdf5/WriteHDF5Test.java:105]: R scripts for read/write verification added

File: src/main/java/org/apache/sysds/runtime/io/hdf5/H5Exception.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysds.runtime.io.hdf5;
+
+import java.io.IOException;
+
+public class H5Exception extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/hdf5/H5Exception.java]: I renamed it to H5 RuntimeException and extend it from DML RuntimeException.

File: src/main/java/org/apache/sysds/runtime/io/WriterHDF5.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+
+public class WriterHDF5 extends MatrixWriter {
+
+	protected static FileFormatPropertiesHDF5 _props = null;
+
+	protected int _replication = -1;
+
+	public static final int BLOCKSIZE_J = 32; //32 cells (typically ~512B, should be less than write buffer of 1KB)
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/WriterHDF5.java]: I changed it.

File: src/main/java/org/apache/sysds/runtime/io/WriterHDF5.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+		try {
+			// HDF5 format don't support spars matrix
+			// Write the data to the datasets.
+			for(int i = rl; i < rlen; i++) {
+				double[] dataRow = new double[clen];
+				//write row chunk-wise to prevent OOM on large number of columns
+				for(int bj = 0; bj < clen; bj += BLOCKSIZE_J) {
+					for(int j = bj; j < Math.min(clen, bj + BLOCKSIZE_J); j++) {
+						double lvalue = src.getValueDenseUnsafe(i, j);
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/WriterHDF5.java]: I changed it.

File: src/main/java/org/apache/sysds/runtime/io/hdf5/H5.java [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+				bb.writeDouble(d);
+			}
+			rootObject.getBufferedOutputStream().write(bb.noOrderBuild().array());
+		}
+		catch(Exception exception) {
+			throw new H5Exception(exception);
+		}
+	}
+
+	public static void H5Dwrite(H5RootObject rootObject, double[][] data) {
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/hdf5/H5.java:189]: Yes, it was just for a test and never used.

File: src/main/java/org/apache/sysds/parser/DataExpression.java [hunk 16]
```diff
@@ -112,6 +112,9 @@
 	public static final String DELIM_NA_STRING_SEP = "\u00b7";
 	// Parameter names relevant to reading/writing delimited index/libsvmv files
 	public static final String LIBSVM_INDEX_DELIM = "indSep";
+
+	// Parameter names relevant to reading/writing dataset name/hdf5 files
+	public static final String HDF5_DATASET_NAME = "dataset";
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/parser/DataExpression.java:117]: The HDF5 format can contain multiple datasets and the datasets can distinguish by name/groups. So, it required the dataset name when we have multiple datasets inside of the HDF5 file. But for the single dataset inside of a file, it is unnecessary. Again on the write side, I already mentioned we are supporting a single dataset, it is unnecessary for us too. Maybe it is important for the end-user!

File: src/main/java/org/apache/sysds/runtime/io/ReaderHDF5.java [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+		long lnnz = 0;
+		H5RootObject rootObject = H5.H5Fopen(bis);
+		H5ContiguousDataset contiguousDataset = H5.H5Dopen(rootObject, datasetName);
+
+		int[] dims = rootObject.getDimensions();
+		int ncol = dims[1];
+
+		DenseBlock denseBlock = dest.getDenseBlock();
+		for(int i = row; i < rlen; i++) {
+			double[] rowData = H5.H5Dread(rootObject, contiguousDataset, i);
```
[the remark this answers is not in the document]
Comment (fathollahzadeh) [reply src/main/java/org/apache/sysds/runtime/io/ReaderHDF5.java]: I changed the approach. I allocated the Array for the first time (before starting for loop) and kept it until the end of the reading.

**Comment by Baunsgaard** [comment]
Thanks for the PR, i have now merged it.
while merging i \@Ignored a few of the write tests, since these took 1 minute, now instead they take 12sec.

**Pull request closed without merging**